pitch.md

```diff
@@ -23,38 +86,122 @@ should be more specific than "The user describes properties of unicorns.", e.g.,
+potential sound. At runtime, it can be synthesized into a sound via the
+computer's DAC, or perhaps it can be written to a file.) A source has no input
+(though it may have _parameters_, see below) and a single output. There will be
+several types of primitive sources, distinguished by the characteristics of the
+signals they produce. Examples include frequency modulation synthesizers and
+physical modeling synthesizers.
+
+An _effect_ is a component which modifies a signal. Its input is a signal, and
+its output is a new, transformed signal. Examples of primitive effects may
+include equalization, compression, and distortion.
+
+Effects can be combined into chains by connecting the output of one effect to
+the input of the next effect. It is easy to see that a chain of effects is
+itself an effect, since it has one input, which is passed through all the
+effects in series, until we reach the final output. Also note that if we
+connect the output of a source to the input of an effect, we get a source.
+
+This describes the basic way in which the user can create new components by
+chaining together existing components. But what if the user wants to modify a
+signal by sending it through two effects chains in parallel, and then combining
+the results? This leads to the third kind of component:
+
+A _multiplexer_ is a component which takes two or more inputs and produces a
+single output. They are distinguished by the method used to combine the inputs
+into a single signal. For example, you could have a multiplexer whose output is
+the sum of its inputs, or one whose output is a weighted average of its inputs.
+
+The last feature I propose for the language is designed to aid in code reuse.
+Components can have parameters which modify the behavior of the component.
+These are not inputs; the feature behaves more like C++ templates in that each
+instantiation of a component with different parameters is conceptually a
+different component, which just happens to use the same code. For example, an
+instrument (which is really just a source) may have parameters to determine the
+pitch and amplitude of the signal it produces.
 
 ## Operation (semantics)
 _What might happen when a program runs? Is there a computational model that
 corresponds with your domain? How does a program interact with the
 user? What kinds of semantic (i.e., non-syntax) errors might occur, and how 
 might they be communicated to the user?_
 
+The computational model is "production rules". A program written in CSound++
+takes as input a MIDI score. This can be a prewritten MIDI file, are a live
+stream of MIDI events from a MIDI controller such as a keyboard. The events
+are then parsed into lists of parameters, including information such as pitch
+and velocity (which generally corresponds to loudness). These parameters are
+then used to instantiate a CSound++ instrument, which produces an audio signal,
+the output of the program.
```

Comment:
I realized I forgot to discuss possible errors that might occur. In critique, we talked about some situations that are not errors but that might merit a warning. For example, it may be possible to chain a bunch of effects together so that they kill any input signal and output a constant, or something like that. If that happens, it is probably not what the user intended, and so a warning would be helpful. However, I think catching situations like that at compile time could be extremely difficult.

In terms of actual errors, this is a language with a type system. So if, for example, the programmer uses an identifier which is bound to an effect in a situation that requires a source, that would be an error. This should prevent the program from being compiled and result in a descriptive error message being displayed for the user.